[Step2]연관 관계 매핑 PR 요청 드립니다.

re-mora opened the pull request:
안녕하세요 :) 교육생 최용석입니다.

2단계 연관 관계 매핑 미션 구현이 완료되어, 코드 리뷰 요청드립니다.

이번 미션 역시 미흡한 부분이 있다면 피드백 부탁드립니다 :)

--- 

이번 PR에 step1에서 피드백 주셨던 JPA Audit 동작 방식에 대한 학습 테스트 TC를 아래 커밋에 작성하였습니다.
  - [JPA Audit 동작 방식 학습을 위한 TC 작성](https://github.com/next-step/jwp-qna/commit/736a080876a814d8bc60a4f43a80ffdcf2cd0240)

말씀하셨던 부분 중 언제 Audit을 사용하지 않는 게 좋을지에 대한 부분을 고민해 보았지만,
아직 경험이 부족하여 예제의 DeleteHistory와 같이 insert 이후 불변성을 가져야 하는 이력성 데이터를 제외하면 사용하지 않아야 하는 경우를 마땅히 떠올리지 못했습니다 😢.. 이 부분에 대해 추가적으로 학습해야 하거나 고민해봐야 하는 내용이 있다면 한번 더 피드백 주시면 감사하겠습니다.

하지만 `JPA Audit 동작 방식`에 대한 키워드를 주신 덕분에 영속성 컨텍스트, flush, 쓰기지연 저장소와 연관지어 JPA Audit의 동작 방식을 조금 더 깊게 살펴 보게되어 많은 공부가 되었습니다.  다시 한번 감사합니다 :)

re-mora committed:
docs: step2 요구사항 분석 및 구현 기능 목록 작성

re-mora committed:
refactor: 질문 엔티티 매핑 수정

질문의 엔티티와 질문 작성자의 연관관계 설정
- `@ManyToOne`을 이용하여 질문 작성자와 질문의 `1:N` 연관관계를 매핑
  - 질문 엔티티 조회 시, 질문 작성자 엔티티의 프록시 초기화 과정에서 발생할 수 있는 `N+1` 문제 방지를 위한 LazyLoading 옵션 설정
- `@JoinColumn`을 이용하여 외래 키 관련 매핑 정보 설정

코드 리팩토링
- 참조 없는 set메소드 제거
- TC 내 명확하지 않은 검증 내용 수정
- TC 작성에 반복되는 하드코딩 부 추출 및 코드 변경에 따른 중복 제거

re-mora committed:
refactor: 질문 엔티티와 질문 작성자의 연관관계 설정 수정

질문의 엔티티와 질문 작성자의 연관관계 설정 시 Not null 설정
- 코드 변경에 따른 TC 수정

re-mora committed:
refactor: 답변 엔티티 매핑 수정

답변과 답변 작성자, 질문과 답변의 다중성에 따른 연관관계 설정
- `@ManyToOne`을 이용하여 `1:N` 연관관계 설정
- *ToOne 관계에서 프록시 초기화 시 발생하는 `N+1` 문제 방지를 위한 LazyLoading 옵션 설정
- `@JoinColumn`을 이용하여 외래 키 관련 매핑 정보 설정
- 답변 작성자와 질문의 연관관계 설정 시, Not null 제약조건 설정

질문 연관관계 설정으로 인해 제거된 기존 속성을 기반으로 하는 쿼리 메소드명 수정
- AS-IS : findByQuestionIdAndDeletedFalse(Long questionId)
- TO-BE : findByQuestionAndDeletedFalse(Question question)

코드 리팩토링
- 참조 없는 set메소드 제거
- TC 내 명확하지 않은 검증 항목 수정
- TC 작성에 반복되는 하드코딩 부 추출 및 코드 변경에 따른 중복 제거

re-mora committed:
refactor: 삭제 이력 엔티티 매핑 수정

삭제 이력과 삭제 요청자의 연관관계 설정
- `@ManyToOne`을 이용하여 `1:N` 연관관계 설정
- *ToOne 관계에서 프록시 초기화 시 발생하는 `N+1` 문제 방지를 위한 LazyLoading 옵션 설정
- `@JoinColumn`을 이용하여 외래 키 관련 매핑 정보 설정
- 삭제 요청자와의 연관관계 설정 시, Not null 제약조건 설정

re-mora committed:
refactor: 사용자 엔티티 수정

참조 없는 set메서드 삭제
Unique 제약조건 명 수정
- AS-IS : USER_ID_UNIQUE
- TO-BE : UK_USER_ID

re-mora committed:
refactor: step1 피드백 JPA Audit 동작 방식 학습테스트 : modifyOnCreate = false 옵션 적용

엔티티가 영속처리 되는 시점에 Modified 관련 이벤트를 처리하지 않도록 EnableJpaAuditing의 modifyOnCreate 옵션 설정
- 설정 변경에 따른 각Entity 생성 TC의 수정일시 검증 부 수정
  - AS-IS : assertThat(entity.getUpdatedAt()).isNotNull()
  - TO-BE : assertThat(entity.getUpdatedAt()).isNull()
- 변경 감지에 의한 Entity 수정 시 수정일시 할당 검증 항목 추가

re-mora committed:
test: step1 피드백 JPA Audit 동작 방식 학습테스트를 위한 TC 작성

- TC 작성을 위한 예제 Entity 및 Repository 작성 : Member
- 하나의 트랜잭션 내에서 여러 엔티티 저장 시, 생성일시 값 할당 확인을 위한 TC 작성
- 하나의 트랜잭션 내에서 여러 엔티티 저장 시, 수정일시 값 할당 확인을 위한 TC 작성
- 신규 Entity 여부를 판별하는 기본 전략 확인 TC 작성
- @Modifying를 이용해 JPQL로 작성된 쿼리 실행 시, JPA Audit 동작 확인을 위한 TC 작성

seokhongkim commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -19,9 +23,21 @@ public class Answer extends BaseEntity {
-    private Long writerId;
-
-    private Long questionId;
+    @ManyToOne(fetch = FetchType.LAZY)
```
연관관계를 설정하실 때 fetchType 을 `FetchType.LAZY` 로 설정하신 이유가 있을까요?

re-mora replied:
*ToOne의 경우 기본으로 적용된 EAGER 설정으로 인해  N+1이 발생하는 할 수 있는 경우를 방지하고자 LazyLoading 설정을 적용하였습니다. 

혹시 이 부분에 대해 제가 이해가 부족한 부분이 있으면 가감없이 피드백 부탁드립니다 :)

seokhongkim replied:
N+1 쿼리 방지 목적으로 설정하신 것 좋습니다.
N+1 쿼리를 방지하는 다른 방법들도 같이 학습해보시면 좋을 것 같네요!

seokhongkim commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -114,15 +122,15 @@ public boolean equals(Object o) {
     }
 
     @Override
     public String toString() {
```
연관관계가 설정된 엔티티의 `toString` 을 구현할 때 주의해야 할 점은 어떤 게 있을까요?

re-mora replied:
양뱡향 연관관계를 가지는 객체들의 toString 구현 시, 순환 참조로 인해 StackOverFlow 문제가 발생할 수 있음에 주의해야 할 것 같습니다. 이 경우 Entity를 별도의 DTO 객체로 변환하거나, 다른쪽의 toString에서 참조를 끊어내는 방법을 고려해 볼 수 있을 것 같아요.

seokhongkim replied:
맞습니다. 말씀하신 것처럼 양방향 참조인 경우 스택 오버플로우가 발생할 수 있습니다.

또한 연관관계 설정된 엔티티가 fetchType 이 LAZY 인 경우, `toString` 이 트랜잭션 바깥에서 호출된다면 엔티티가 없을 때 LazyInitializationException 이 발생할 수도 있습니다.
어느곳에서 사용될 지 모르므로 이런 점을 고려하여 `toString` 을 구현해주시면 좋을 것 같습니다. :)

seokhongkim commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -19,9 +23,21 @@ public class Answer extends BaseEntity {
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(
+        name = "writer_id",
+        foreignKey = @ForeignKey(name = "FK_ANSWER_WRITER"),
```
명시적인 FK 설정 좋습니다! 👍

seokhongkim commented on src/main/java/qna/service/QnaService.java:
```diff
@@ -39,7 +39,7 @@ public void deleteQuestion(User loginUser, Long questionId) throws CannotDeleteE
         }
 
-        List<Answer> answers = answerRepository.findByQuestionIdAndDeletedFalse(questionId);
+        List<Answer> answers = answerRepository.findByQuestionAndDeletedFalse(question);
```
이제 `Answer` 와 `Question` 은 연관관계가 맺어졌으니, `Question` 에게 삭제되지 않은 `Answer` 목록을 달라고 요청하는 방향으로 리팩토링해보면 어떨까요?

미션의 힌트에 아래와 같은 내용이 있네요!
```java
Question question = findQuestionById(questionId);
List<Answer> answers = question.getAnswers();
```

re-mora replied:
꼼꼼히 살펴봐야 했는데 이 부분을 놓쳤네요!
쿼리메소드를 이용해 객체를 조회하는 부분을 연관관계를 가지는 질문 객체를 통해 접근하도록 수정 하겠습니다 :)

꼼꼼하게 살펴봐주셔서 감사합니다 👍

seokhongkim commented on src/test/java/qna/audit/AuditingTest.java:
```diff
@@ -0,0 +1,125 @@
+@DataJpaTest
+@TestConstructor(autowireMode = AutowireMode.ALL)
+@DisplayName("Test:JPA Auditing")
+class AuditingTest {
```
학습을 위한 테스트 좋습니다! 💯

choi-ys replied:
> 

ㅇ

seokhongkim reviewed (changes requested):
용석님, JPA 2단계 미션 잘 구현해주셨습니다! 👍 

지난번에 말씀드린 JPA Audit 에 대해 굉장히 많이 학습하신 것 같고, 학습 테스트도 작성해보신 점이 인상적이었습니다.
이번 미션을 작은 작업 단위로 나눠서 잘 진행해주셨고, 커밋도 작게 나눠서 해주신 점도 좋았습니다.

Audit 을 사용하지 않는 게 좋을지 고민해보시고 질문주셔서, 이에 대해 조금 더 생각해볼 수 있는 구체적인 상황 예시를 말씀 드릴게요!

임의의 쇼핑몰 시스템이 있을 때, 이 시스템은 주문이 발생하면 주문, 주문한 상품, 결제 정보 엔티티를 저장합니다.
이때 주문, 주문한 상품, 결제 정보의 생성 일시가 동일해야 할까요? 다른 값이어도 괜찮을까요?
시스템의 규모가 점점 커지며 샤딩 또는 파티셔닝을 고려하게 되고, 이때 생성 일시가 기준이 될 수도 있다면 고민해 볼 필요가 있을 것 같습니다.
또한, 데이터를 분석하기 위해 주기적으로 별도의 시스템으로 데이터를 옮긴다면, 데이터를 옮기기 위한 기준 시간이 생성 또는 수정 일시가 될 수 있을 것 같습니다. 이런 상황에서 Audit 을 사용한다면 함께 옮겨져야 할 데이터가 같이 옮겨지지 않는 상황이 발생할 수 있을 것 같습니다.
따라서 Audit 을 사용할 때는 코드의 편의성 외에 전체 시스템의 구조, 데이터 사용 방법 등을 포함해서 고민해 볼 필요가 있다고 생각합니다.

코드 작업해주신 부분에 대해서는 조금 고민해보실 수 있는 내용과 추가로 진행해주시면 좋을 작업 위주로 코멘트를 조금 남겼습니다.

확인 부탁드립니다. 🙏

re-mora committed:
refactor: step2 피드백 양방향 연관관계 적용 : Question <-> Answer

질문과 답변의 양방향 연관관계 설정
- 양방향 연관관계 설정 시, 질문 엔티티를 통해 외래키가 관리 되지 않도록 mappedBy 속성 명시
- 객체 시점에 값 동기화를 위한 연관관계 편의 메소드 설정

양방향 연관관계 적용
- QnaService의 질문 삭제 기능에서 미삭제 답변 목록 조회 시, 쿼리 호출 대신 조회된 질문에서 접근 하도록 수정

로직 수정으로 인해 깨지는 TC 수정
- 불필요 스터빙 구문 및 참조가 없는 mock 객체 제거

re-mora commented:
> 용석님, JPA 2단계 미션 잘 구현해주셨습니다! 👍
> 
> 지난번에 말씀드린 JPA Audit 에 대해 굉장히 많이 학습하신 것 같고, 학습 테스트도 작성해보신 점이 인상적이었습니다. 이번 미션을 작은 작업 단위로 나눠서 잘 진행해주셨고, 커밋도 작게 나눠서 해주신 점도 좋았습니다.
> 
> Audit 을 사용하지 않는 게 좋을지 고민해보시고 질문주셔서, 이에 대해 조금 더 생각해볼 수 있는 구체적인 상황 예시를 말씀 드릴게요!
> 
> 임의의 쇼핑몰 시스템이 있을 때, 이 시스템은 주문이 발생하면 주문, 주문한 상품, 결제 정보 엔티티를 저장합니다. 이때 주문, 주문한 상품, 결제 정보의 생성 일시가 동일해야 할까요? 다른 값이어도 괜찮을까요? 시스템의 규모가 점점 커지며 샤딩 또는 파티셔닝을 고려하게 되고, 이때 생성 일시가 기준이 될 수도 있다면 고민해 볼 필요가 있을 것 같습니다. 또한, 데이터를 분석하기 위해 주기적으로 별도의 시스템으로 데이터를 옮긴다면, 데이터를 옮기기 위한 기준 시간이 생성 또는 수정 일시가 될 수 있을 것 같습니다. 이런 상황에서 Audit 을 사용한다면 함께 옮겨져야 할 데이터가 같이 옮겨지지 않는 상황이 발생할 수 있을 것 같습니다. 따라서 Audit 을 사용할 때는 코드의 편의성 외에 전체 시스템의 구조, 데이터 사용 방법 등을 포함해서 고민해 볼 필요가 있다고 생각합니다.
> 
> 코드 작업해주신 부분에 대해서는 조금 고민해보실 수 있는 내용과 추가로 진행해주시면 좋을 작업 위주로 코멘트를 조금 남겼습니다.
> 
> 확인 부탁드립니다. 🙏

구체적인 예시를 들어주신 덕분에, 말씀하셨던 JPA Audit의 사용여부에 대해 고민했던 부분이 한번에 와닿았습니다. 
특히 마지막의 `따라서 Audit 을 사용할 때는 코드의 편의성 외에 전체 시스템의 구조, 데이터 사용 방법 등을 포함해서 고민해 볼 필요가 있다고 생각합니다.` 내용은 Audit 뿐만 아니라 다양한 경우에도 말씀하신 관점에서 고민한다면 코드와 아키텍처를 같이 고려하고 시야를 조금 더 넓게 바라볼 수 있을 것 같습니다.
 
자세하게 설명해주셔서 너무 감사드립니다! 👍

seokhongkim commented on src/main/java/qna/service/QnaService.java:
```diff
@@ -39,7 +39,7 @@ public void deleteQuestion(User loginUser, Long questionId) throws CannotDeleteE
         }
 
-        List<Answer> answers = answerRepository.findByQuestionIdAndDeletedFalse(questionId);
+        List<Answer> answers = question.getAnswers();
```
1. `AnswerRepository` 는 더 이상 직접 사용하지 않으므로 의존성을 삭제할 수 있을 것 같아요. 
2. findByQuestionAndDeletedFalse` 메소드도 사용처가 없다면 삭제할 수 있을 것 같습니다.
3. 리팩토링 과정에서 기존에는 `삭제되지 않은 답변` 을 가져왔는데, `모든 답변` 을 가져오도록 로직도 변경된 것 같습니다.

seokhongkim reviewed (approved):
용석님, JPA 2단계 미션 피드백 잘 반영해주셨습니다. 👍 

코멘트를 하나 더 남겼는데, 이 부분은 3단계 미션 진행하면서 보셔도 괜찮을 것 같습니다.
2단계 PR은 이만 머지하도록 하겠습니다.

3단계 미션에서 뵙겠습니다. 😀
